### PR TITLE
Fix pointer types and use `std::mem` on `std::tx`

### PR DESCRIPTION
**Pull request by AlicanC**
Resolves https://github.com/FuelLabs/sway/issues/2015

---

[Make all pointers u64](https://github.com/FuelLabs/sway/pull/2208/commits/b10f54a75e51cd3df17a53d9d424211dab3ade11)

Some were `u32`s. Also checked outside `std::tx` for more and couldn't see any.

[Suffix pointer returning functions with _pointer](https://github.com/FuelLabs/sway/pull/2208/commits/b29f78daa57dd3df6490cea4d16b65a8869d6942)

Some were using `_offset`. `_ptr` is also an option but we use `_pointer` outside `std::tx` too. 

[Replace assembly with internals](https://github.com/FuelLabs/sway/pull/2208/commits/acd56a4318ea72fc00039e065a05626d87df7fa3)

Used `std::mem` and `std::tx` internals instead of asm. There were more opportunities but I think they'd be better done in a separate PR. (Probably https://github.com/FuelLabs/sway/issues/2146)

### CODE REVIEW

File: sway-lib-std/src/tx.sw [hunk 1]
```diff
@@ -144,67 +144,52 @@ pub fn tx_receipts_root() -> b256 {
...
-            r1: T
-        }
-    } else {
-        asm(r1: ptr) {
-            lw r1 r1 i0;
-            r1: T
-        }
-    }
+    let ptr = tx_script_data_start_pointer();
+    read(ptr)
```
Comment (mohammadfawaz) [inline sway-lib-std/src/tx.sw:168]:
```suggestion
    read(tx_script_data_start_pointer())
```

File: sway-lib-std/src/tx.sw [hunk 2]
```diff
@@ -144,67 +144,52 @@ pub fn tx_receipts_root() -> b256 {
...
 /// Get the script bytecode
 /// Must be cast to a u64 array, with sufficient length to contain the bytecode.
 /// Bytecode will be padded to next whole word.
 pub fn tx_script_bytecode<T>() -> T {
-    let script_ptr = tx_script_start_offset();
-    asm(r1: script_ptr) {
-        r1: T
-    }
+    let script_ptr = tx_script_start_pointer();
+    read(script_ptr)
```
Comment (mohammadfawaz) [inline sway-lib-std/src/tx.sw:176]:
```suggestion
    read(tx_script_start_pointer())
```

File: sway-lib-std/src/tx.sw [hunk 3]
```diff
@@ -261,7 +246,7 @@ pub fn tx_predicate_data_start_offset() -> u64 {
 }
 
 pub fn get_predicate_data<T>() -> T {
-    let ptr = tx_predicate_data_start_offset();
+    let ptr = tx_predicate_data_start_pointer();
     read(ptr)
```
Comment (mohammadfawaz) [inline sway-lib-std/src/tx.sw:250]:
```suggestion
    read(tx_predicate_data_start_pointer())
```

**Review by mohammadfawaz — COMMENTED** [review]
Nice! Just a few nits.

**Review by mohammadfawaz — DISMISSED** [review]

**Review by simonr0204 — DISMISSED** [review]
LGTM

**Review by adlerjohn — APPROVED** [review]

**Review by mohammadfawaz — APPROVED** [review]

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**